crypto/totp_test.go

```diff
@@ -0,0 +1,242 @@
+package crypto
+
+import (
+    "crypto/sha1"
+    "encoding/base32"
+    assert1 "github.com/stretchr/testify/assert"
+    "os"
+    "testing"
+    "time"
+)
+
+func TestDynamicTruncate(t *testing.T) {
+    type args struct {
+        hmacResult []byte
+        digits     int
+    }
+    tests := []struct {
+        name string
+        args args
+        want string
+    }{
+        {
+            "test from RFC",
+            args{[]byte{0x1f, 0x86, 0x98, 0x69, 0x0e, 0x02, 0xca, 0x16, 0x61, 0x85, 0x50, 0xef, 0x7f, 0x19, 0xda, 0x8e, 0x94, 0x5b, 0x55, 0x5a}, 6},
+            "872921",
+        },
+    }
+    for _, tt := range tests {
+        t.Run(tt.name, func(t *testing.T) {
+            if got := DynamicTruncate(tt.args.hmacResult, tt.args.digits); got != tt.want {
+                t.Errorf("DynamicTruncate() = %v, want %v", got, tt.want)
+            }
+        })
+    }
+}
+
+func TestHOTP(t *testing.T) {
```

Comment:
My understanding of HOTP is that the token and server can get out of sync with respect to the counter. This is because the token will increment it's counter every time the user asks to generate a new challenge even if that challenge is never passed to the server resulting in an incremented counter on the server on success. The way this is handled is to allow for some variance in the counter on the server side. I don't see any tests that verify a fail or success on an invalid counter or anything that allows for the counter variance. Is this necessary in your opinion?

Reply:
I am not doing a 'WithVariance' function for HOTP because it is up to the implementation to decide if the validation that is occurring should advance the counter or not. The caller can check the variance by just adding or subtracting one to the counter on failure.

Reply:
LGTM. It didn't occur to me that the TOTP is just an HOTP with a time based variance vs. an absolute counter.